### PR TITLE
Make BlockEncode constructor JIT-friendly

### PR DESCRIPTION
**Pull request by dime10**
The new BlockEncode operator is currently not JIT friendly due to its constructor (whereas the matrix method already seems to be).

Conditionals are replaced with equivalent mathematical functions.

[sc-54971]

### CODE REVIEW

File: pennylane/ops/qubit/matrix_ops.py [hunk 1]
```diff
@@ -519,16 +519,17 @@ def __init__(self, A, wires, id=None):
 
         wires = Wires(wires)
         if pnp.sum(shape_a) <= 2:
-            normalization = A if A > 1 else 1
+            normalization = A
```
[the remark this answers is not in the document]
Comment (DSGuala) [reply pennylane/ops/qubit/matrix_ops.py:522]: This is a mistake :bug: It's intended to only catch scalar inputs, not size-2 vectors. They should be caught even when formatted differently `e.g. [1], [[2]], or [[[[[4]]]]]`

In fact, BlockEncode gives an error for `A=[1,0]`
This shouldn't have gotten through, but we didn't have a test for e.g. `[1,0]` :disappointed: 

We can replace `A if A > 1 else 1` with `qml.math.maximum(A, 1.0)` :+1:

[the remark this answers is not in the document]
Comment (DSGuala) [reply pennylane/ops/qubit/matrix_ops.py]: Pushed a commit that should help with this. Please let me know your thoughts :pray:

File: pennylane/ops/qubit/matrix_ops.py [hunk 2]
```diff
@@ -513,22 +515,27 @@ class BlockEncode(Operation):
...
     def __init__(self, A, wires, id=None):
+        wires = Wires(wires)
         shape_a = qml.math.shape(A)
         if shape_a == () or all(x == 1 for x in shape_a):
             A = qml.math.reshape(A, [1, 1])
-
-        wires = Wires(wires)
-        if pnp.sum(shape_a) <= 2:
-            normalization = A if A > 1 else 1
+            normalization = qml.math.abs(A)
```
Comment (maliasadi) [inline pennylane/ops/qubit/matrix_ops.py:522]: `abs` and other `qml.math` methods can be imported similar to those imported methods in line 24 for consistency.

**Review by maliasadi — COMMENTED** [review]
Thanks @dime10! :raised_hands:

File: tests/ops/qubit/test_matrix_ops.py [hunk 3]
```diff
@@ -963,7 +964,7 @@ def test_blockencode_torch(self, input_matrix, wires, output_matrix):
         ("input_matrix", "wires", "output_matrix"),
         [
             (1, 0, [[1, 0], [0, -1]]),
-            (0.3, 0, [[0.3, 0.9539392], [0.9539392, -0.3]]),
+            ([0.3], 0, [[0.3, 0.9539392], [0.9539392, -0.3]]),
```
Comment (dime10) [inline tests/ops/qubit/test_matrix_ops.py:967]: I changed this input to ensure the now three branches in the constructor are hit.

File: pennylane/ops/qubit/matrix_ops.py [hunk 4]
```diff
@@ -515,22 +515,27 @@ class BlockEncode(Operation):
...
+        elif len(shape_a) == 1:
+            A = qml.math.reshape(A, [1, len(A)])
+            normalization = qml.math.max(A)
+            subspace = (1, shape_a[0], 2 ** len(wires))
+
         else:
-            normalization = max(
+            normalization = qml.math.maximum(
                 norm(A @ qml.math.transpose(qml.math.conj(A)), ord=pnp.inf),
                 norm(qml.math.transpose(qml.math.conj(A)) @ A, ord=pnp.inf),
```
Comment (Jaybsoni) [inline pennylane/ops/qubit/matrix_ops.py:533]: Why separate these two cases? We should be able to treat the vector case as a matrix itself, the normalization defined below [ max(norm(A@A^), norm(A^@A)) ]

**Review by Jaybsoni — COMMENTED** [review]
Looks good, just one question to address and then I am happy to approve :)

File: pennylane/ops/qubit/matrix_ops.py (hunk 4 above)
Comment (dime10) [reply pennylane/ops/qubit/matrix_ops.py:533]: This is @DSGuala design, I just made it traceable 😅

Comment (DSGuala) [reply pennylane/ops/qubit/matrix_ops.py:533]: The normalization below can't handle a simple list, e.g. `[0.1,0.2]` because ``norm`` doesn't like being called on a scalar. In the next commit, I changed things around so we're reshaping these one dimensional lists into matrices inside the else statement.

File: pennylane/ops/qubit/matrix_ops.py (hunk 2 above)
Comment (Jaybsoni) [reply pennylane/ops/qubit/matrix_ops.py:522]: +1

**Review by Jaybsoni — APPROVED** [review]
LGTM, sorry for slowing this one down 💯

File: tests/ops/qubit/test_matrix_ops.py [hunk 5]
```diff
@@ -983,13 +984,22 @@ def test_blockencode_torch(self, input_matrix, wires, output_matrix):
...
         assert np.allclose(qml.matrix(op), output_matrix)
         assert qml.math.get_interface(qml.matrix(op)) == "jax"
 
+        # Test jitting behaviour as well.
+        @jax.jit
+        def f(A):
+            op = qml.BlockEncode(A, wires)
+            return qml.matrix(op)
+
+        assert np.allclose(f(input_matrix), output_matrix)
```
Comment (Jaybsoni) [inline tests/ops/qubit/test_matrix_ops.py:1001]: Does this test that it was jitted correctly? What if it didn't jit anything under the hood and just used the default jax execution? 

Asking for my own curiosity, how does one usually test that things are actually being jitted?

Comment (dime10) [reply tests/ops/qubit/test_matrix_ops.py:1001]: Hmm, good question. Maybe we should just trust that if `jax.jit` is applied to the function it does indeed jit 😅
